src/transform/src/normalize_lets.rs

```diff
@@ -351,121 +351,181 @@ mod let_motion {
         // First, promote all `LetRec` nodes above all other nodes.
         let mut worklist = vec![&mut *expr];
         while let Some(expr) = worklist.pop() {
-            digest_lets(expr);
-            if let MirRelationExpr::LetRec {
-                ids: _,
-                values,
-                limits: _,
-                body,
-            } = expr
-            {
-                // The order may not be important, but let's not risk it.
-                worklist.push(body);
+            hoist_bindings(expr);
+            if let MirRelationExpr::LetRec { values, .. } = expr {
                 worklist.extend(values.iter_mut().rev());
             }
         }
+
         // Harvest any potential `Let` nodes, via a post-order traversal.
         post_order_harvest_lets(expr);
     }
 
-    /// Performs a post-order traversal of the `LetRec` nodes at the root of an expression.
+    /// A stand in for the types of bindings we might encounter.
     ///
-    /// The traversal is only of the `LetRec` nodes, for which fear of stack exhaustion is nominal.
-    fn post_order_harvest_lets(expr: &mut MirRelationExpr) {
-        if let MirRelationExpr::LetRec {
-            ids,
-            values,
-            limits,
-            body,
-        } = expr
-        {
-            // Only recursively descend through `LetRec` stages.
-            for value in values.iter_mut() {
-                post_order_harvest_lets(value);
+    /// As we dissolve various `Let` and `LetRec` expressions, a `Binding` will carry
+    /// the relevant information as we hoist it to the root of the expression.
+    enum Binding {
+        // Binding resulting from a `Let` expression.
+        Let(LocalId, MirRelationExpr),
+        // Bindings resulting from a `LetRec` expression.
+        LetRec(Vec<(LocalId, MirRelationExpr, Option<LetRecLimit>)>),
+    }
+
+    /// Hoist all exposed bindings to the root of the expression.
+    ///
+    /// A binding is "exposed" if the path from the root does not cross a LetRec binding.
+    /// After the call, the expression should be a linear sequence of bindings, where each
+    /// `Let` binding is of a let-free expression. There may be `LetRec` expressions in the
+    /// sequence, and their bindings will have hoisted bindings to their root, but not out
+    /// of the binding.
+    fn hoist_bindings(expr: &mut MirRelationExpr) {
+        // Bindings we have extracted but not fully processed.
```

Comment:
Document these in terms of binder order: `worklist` is outermost binders to innermost, and `finished` is innermost to outermost.